File: Demos/10. StartedServicesDemo/SimpleServiceDemo/app/src/main/java/org/feup/apm/simpleservice/MainActivity.java
package org.feup.apm.simpleservice;

import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.os.Bundle;
import androidx.appcompat.app.AppCompatActivity;
import android.view.View;
import android.widget.RadioGroup;
import android.widget.ScrollView;
import android.widget.TextView;

public class MainActivity extends AppCompatActivity implements RadioGroup.OnCheckedChangeListener {
  static MainActivity myInstance = null;

  private int counter = 1;
  private RadioGroup rg;
  private TextView console;
  private ScrollView scroll;
  private BroadcastReceiver receiver;

  @Override
  public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    setContentView(R.layout.main);
    rg = (RadioGroup) findViewById(R.id.radioGroup1);
    rg.setOnCheckedChangeListener(this);
    console = (TextView) findViewById(R.id.textView);
    scroll = (ScrollView) findViewById(R.id.scView);
    MainActivity.myInstance = this;
    receiver = new BroadcastReceiver() {
      @Override
      public void onReceive(Context context, Intent intent) {
        String message = intent.getStringExtra("message");
        addText(message);
      }
    };
  }

  @Override
  protected void onResume() {
    super.onResume();
    IntentFilter intentFilter = new IntentFilter("android.intent.action.RCV_MESS");
    registerReceiver(receiver, intentFilter);
  }

  @Override
  protected void onPause() {
    super.onPause();
    unregisterReceiver(receiver);
  }

  public void addText(String text) {
    String old = console.getText().toString();
    console.setText(old + text + "\n");
    scroll.fullScroll(ScrollView.FOCUS_DOWN);
  }

  public void doButtonClick(View view) {
    Intent intent;
    switch(view.getId()) {
      case R.id.startBtn:
        addText("Call service #" + counter + ", tid = " + android.os.Process.myTid());
        if (rg.getCheckedRadioButtonId() == R.id.radio0)
          intent = new Intent(this, BackgroundService.class);
        else
          intent = new Intent("org.feup.apm.intents.ACTION_TESTSERVICE");
        intent.putExtra("counter", counter++);
        startService(intent);
        break;
      case R.id.stopBtn:
        stopService(rg.getCheckedRadioButtonId());
        break;
      case R.id.clearBtn:
        console.setText("");
        scroll.fullScroll(ScrollView.FOCUS_UP);
        break;
    }
  }

  public void onCheckedChanged(RadioGroup rgr, int id) {
    if (id == R.id.radio0)
      stopService(R.id.radio1);
    else
      stopService(R.id.radio0);
  }

  private void stopService(int id) {
    Intent intent;

  	if (id == R.id.radio0)
      intent = new Intent(this, BackgroundService.class);
  	else
  	  intent = new Intent("org.feup.apm.intents.ACTION_TESTSERVICE");
    addText("Call stop service");
  	stopService(intent);
  }

  @Override
  public void onDestroy() {
  	stopService(rg.getCheckedRadioButtonId());
    super.onDestroy();
    MainActivity.myInstance = null;
  }
}

File: Assignments/CodesScanAndGeneration/ZXCodeGenerator/app/src/main/java/org/feup/apm/zxcodegenerator/MainActivity.kt
package org.feup.apm.zxcodegenerator

import android.content.Intent
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.view.View
import android.widget.Button
import android.widget.EditText

class MainActivity : AppCompatActivity() {
  val ed_number by lazy { findViewById<EditText>(R.id.ed_number) }
  val ed_msg by lazy { findViewById<EditText>(R.id.ed_msg) }

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    setContentView(R.layout.activity_main)
    findViewById<Button>(R.id.bt_bar).setOnClickListener { vw -> onButtonClick(vw) }
    findViewById<Button>(R.id.bt_qr).setOnClickListener { vw -> onButtonClick(vw) }
  }

  private fun onButtonClick(vw: View) {
    val intent = Intent(this, ShowCodeActivity::class.java)
    when (vw.id) {
      R.id.bt_bar -> { intent.putExtra("type", 0)
                       val nr = if (ed_number.text.isEmpty())
                         "00000000001"
                       else
                         ed_number.text.toString()
                       intent.putExtra("value", nr) }
      R.id.bt_qr -> { intent.putExtra("type", 1)
                      val msg = if (ed_msg.text.isEmpty())
                        "Just a simple text message"
                      else
                        ed_msg.text.toString()
                      intent.putExtra("value", msg) }
    }
    startActivity(intent)
  }
}